### ISSUE TITLE
LSTMLayer does not work when a `SharedTensorVariable` is passed for `hid_init`

### ISSUE THREAD

**Issue by riklopfer** [Lasagne/Lasagne#542]
Motivated by this thread: https://groups.google.com/forum/#!topic/lasagne-users/YUt6Bo4OFTA

Currently, it will only work with `TensorVariable` and `Initializer`.

**Comment by skaae** [comment]
As far as I know we are removing the support for `hid_init` being a `TensorVariable`,  please see https://github.com/Lasagne/Lasagne/issues/462 .

Instead `hid_init` should be a `Initializer` or an `Layer`. We came to the conclusion that the implementation would be cleaner and we still would support the same usecases. Do you have any usescases that will not be supported when `TensorVariable` is removed?

**Comment by riklopfer** [comment]
> Do you have any usescases that will not be supported when TensorVariable is removed?
Nope. 

Will https://github.com/Lasagne/Lasagne/issues/462 allow for variable batch sizes when `hid_init` is  update to the previous output?

**Comment by skaae** [comment]
Currently we store a (1, num_units) vector which is then expaned to be `batch_size, num_units`.
I guess for the InputLayer case it make mosts sense to expect the input to be  `batch_size, num_units` e.g. in a Encoder/Decoder structure. 

In anycase variable batchsize will be supported.

**Comment by f0k** [comment]
> Do you have any usescases that will not be supported when `TensorVariable` is removed?

Note that `TensorVariable` will not actually be removed, just the special-case handling in the recurrent layers. Due to #11, any layer parameter can be a `TensorVariable` already. The difference will be that after removing the special case, `hid_init` being a `TensorVariable` would be treated the same as `hid_init` being a `SharedVariable` or `ndarray` or `callable`, i.e., it will be a (potentially learnable) parameter replicated over the batch.

If you remove the special case for `TensorVariable`, a `TensorSharedVariable` will work fine, but it will also be suspect to replication over the batch.

Instead of #543, you should remove the special case for `TensorVariable` and add a parameter that controls whether `hid_init` is to be replicated over the batch or not.

**Comment by skaae** [comment]
> Instead of #543, you should remove the special case for TensorVariable and add a parameter that controls whether hid_init is to be replicated over the batch or not.

Yes, we should defenitely support the input not being replicated.

**Comment by skaae** [comment]
This can be closed when https://github.com/Lasagne/Lasagne/pull/549 is merged.

**Comment by f0k** [comment]
> This can be closed when #549 is merged.

No, we still need an option for not replicating the state over the batch size to cover the use case by @riklopfer. With #549, when you pass a SharedVariable, the recurrent layer will assume it's the initial state for a single sequence, not for a batch of sequences. `add_param()` will then fail because the shared variable passed by @riklopfer has too many dimensions.

**Comment by skaae** [comment]
For @riklopfer use case couldn't you use an input layer for hid_init? 
Then you can manually pass in zeros or the previous hidden states? 
You would have to make both matrices ``[batch_size x num_inputs]``

That would  fail if you want to learn the initial states in the first batch in 
a language model but i think that is a non-important corner case?

Sorry for the late reply.

**Comment by f0k** [comment]
> For @riklopfer use case couldn't you use an input layer for hid_init?

Yes, you could even wrap your shared variable (of size `batch_size x num_units`) in an `InputLayer`. In some sense, this may even be more appropriate since it's not meant to be a parameter of the model then, just an input.

Regardless of this, when we address #429, we can add another parameter like `keep_state` to directly address @riklopfer's use case of not resetting the hidden state between calls. This would create a `batch_size x num_units` shared variable for the initial hidden state and construct the necessary updates in `get_output_for()` to always remember the final hidden state and use it as the initial state for the next call. We don't really need a separate parameter like `hid_init_per_sequence` that disables broadcasting/replication of states over the batch.
